sdk/src/azure/core/az_mqtt5_request_hfsm.c

```diff
@@ -0,0 +1,681 @@
+            this_policy->_internal.correlation_id);
+        _az_RETURN_IF_FAILED(_az_hfsm_transition_peer((_az_hfsm*)me, waiting, completed));
+      }
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_REQUEST_FAULTED:
+    {
+      az_span* correlation_id = (az_span*)event.data;
+      if (az_span_is_content_equal(*correlation_id, this_policy->_internal.correlation_id))
+      {
+        log_event(
+            event,
+            AZ_SPAN_FROM_STR("az_mqtt5_request_policy/waiting"),
+            this_policy->_internal.correlation_id);
+        _az_RETURN_IF_FAILED(_az_hfsm_transition_peer((_az_hfsm*)me, waiting, faulted));
+      }
+      break;
+    }
+
+    case AZ_HFSM_EVENT_TIMEOUT:
+    {
+      if (event.data == &this_policy->_internal.request_completion_timer)
+      {
+        log_event(
+            event,
+            AZ_SPAN_FROM_STR("az_mqtt5_request_policy/waiting"),
+            this_policy->_internal.correlation_id);
+        // if execution times out, send failure to application and go to faulted
+        az_mqtt5_rpc_client_rsp_event_data resp_data
+            = { .response_payload = AZ_SPAN_EMPTY,
+                .status = AZ_MQTT5_RPC_STATUS_TIMEOUT,
+                .error_message = AZ_SPAN_FROM_STR("Execution timed out."),
+                .content_type = AZ_SPAN_EMPTY,
+                .correlation_id = this_policy->_internal.correlation_id };
+
+        _az_RETURN_IF_FAILED(_az_mqtt5_connection_api_callback(
+            this_policy->_internal.connection,
+            (az_event){ .type = AZ_MQTT5_EVENT_RPC_CLIENT_ERROR_RSP, .data = &resp_data }));
+
+        _az_RETURN_IF_FAILED(_az_hfsm_transition_peer((_az_hfsm*)me, waiting, faulted));
+      }
+      break;
+    }
+
+    default:
+      log_event(
+          event,
+          AZ_SPAN_FROM_STR("az_mqtt5_request_policy/waiting"),
+          this_policy->_internal.correlation_id);
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+static az_result completed(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+  az_mqtt5_request* this_policy = (az_mqtt5_request*)me;
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+      log_event(
+          event,
+          AZ_SPAN_FROM_STR("az_mqtt5_request_policy/completed"),
+          this_policy->_internal.correlation_id);
+      _request_stop_timer(&this_policy->_internal.request_completion_timer);
+      break;
+
+    case AZ_HFSM_EVENT_EXIT:
+      log_event(
+          event,
+          AZ_SPAN_FROM_STR("az_mqtt5_request_policy/completed"),
+          this_policy->_internal.correlation_id);
+      break;
+
+    case AZ_MQTT5_EVENT_PUBACK_RSP:
+    case AZ_MQTT5_EVENT_REQUEST_INIT:
+    case AZ_MQTT5_EVENT_REQUEST_COMPLETE:
+    case AZ_MQTT5_EVENT_REQUEST_FAULTED:
+    case AZ_HFSM_EVENT_TIMEOUT:
+      // ignore
+      break;
+
+    default:
+      log_event(
+          event,
+          AZ_SPAN_FROM_STR("az_mqtt5_request_policy/completed"),
```

Comment:
Hmm interesting. This would require some more parsing of the events to make sure they are for this request, but I think that might make sense. I'll take a look at it

Reply:
I updated this across all of the states to make sure these events send invalid_state where logical